assets/js/components/projects/ProjectSettings.jsx

```diff
@@ -0,0 +1,348 @@
+
+class ProjectSettings extends Component {
+  componentDidMount() {
+    if (!this.props.isLoading) {
+      this.setStateFromProject()
+    }
+  }
+
+  componentDidUpdate(prevProps) {
+    if (prevProps.isLoading && !this.props.isLoading) {
+      this.setStateFromProject()
+    }
+    if (prevProps.isSaving && !this.props.isSaving && Object.keys(this.props.errors).length == 0) {
+      this.initialState = JSON.parse(JSON.stringify(this.state))
+      this.forceUpdate() // because we didn't touch the state
+    }
+  }
+
+  setStateFromProject() {
+    const { project } = this.props
+    var state = {
+      name: project.name,
+      timezone: project.timezone,
+      colourScheme: project.colourScheme,
+      initialSuccessRate: project.initialSuccessRate || "",
+      eligibilityRate: project.eligibilityRate || "",
+      responseRate: project.responseRate || "",
+      validRespondentRate: project.validRespondentRate || "",
+      detailedRates: project.eligibilityRate != null,
+      archiveAction: project.readOnly ? "unarchive" : "archive",
+    }
+    this.initialState = JSON.parse(JSON.stringify(state))
+    this.setState(state)
+  }
+
+  toggleDetailedRates() {
+    const { detailedRates } = this.state
+    this.setState({ detailedRates: !detailedRates })
+  }
+
+  saveProjectSettings() {
+    const { dispatch, project } = this.props
+    const changes = merge({}, project, {
+      name: this.state.name,
+      timezone: this.state.timezone,
+      colourScheme: this.state.colourScheme,
+      initialSuccessRate: parseFloat(this.state.initialSuccessRate, 10) || null,
+      eligibilityRate: parseFloat(this.state.eligibilityRate, 10) || null,
+      responseRate: parseFloat(this.state.responseRate, 10) || null,
+      validRespondentRate: parseFloat(this.state.validRespondentRate, 10) || null,
+    })
+    dispatch(projectActions.updateProject(changes))
+  }
+
+  updateInitialSuccessRate() {
+    const { eligibilityRate, responseRate, validRespondentRate } = this.state
+
+    if (eligibilityRate && responseRate && validRespondentRate) {
+      const isr = (eligibilityRate * responseRate * validRespondentRate).toFixed(4)
+      this.setState({ initialSuccessRate: isr })
+    } else {
+      this.setState({ initialSuccessRate: "" })
+    }
+  }
+
+  archiveOrUnarchive(archived: boolean) {
+    const { dispatch, project } = this.state
+    const { archiveAction } = this.state
+    dispatch(actions.archiveOrUnarchive(project, archiveAction))
+    this.setState({ archiveAction: archiveAction == "archive" ? "unarchive" : "archive" })
+  }
+
+  spanErrors(field) {
+    const { errors } = this.props
+
+    if (errors[field]) {
+      return (
+        <span className="error">
+          {errors[field].map((error) => (
+            <div key={error}>{error}</div>
+          ))}
+        </span>
+      )
+    } else {
+      return null
+    }
+  }
+
+  updateRate(newRate) {
+    this.setState(newRate, () => this.updateInitialSuccessRate())
```

Comment:
`setState()` is asynchronous, so we must call `updateInitialSuccessRate()` in a callback to make sure that it runs after the state has been updated, otherwise it would run _before_ the state is updated (oops).